Rectified sys.database_principals view which was missing SID

ayushdsh opened the pull request:
### Description

sys.database_principals.sid acts as a foreign key and links to sys.server_principals.sid This is how SQL users are linked to their respective logins. Certain database_principals like db roles will not have a corresponding mapping in server_principals, hence they have a SID which is generated by default by sql server at creation. In this view, we want to map the users (the defaults like dbo as well) to the server_principals. The owner of the database is the login which is linked to 'dbo'. Hence, we will be mapping the owner of the current db to 'dbo'. Other user-defined logins are already part of the view currently.

Also updated the sys.user_token view to include the SIDs.

Differences between SQL Server and BBF - 

1. Orphaned users still show the SID of the deleted login in SQL server, however in BBF since we are performing joins on other system catalogs, BBF outputs NULL
2. DB roles (fixed and user-defined) do not have any mapping of SID in server_principals; the SIDs given to db roles are by default generated by sql server at creation. Hence, in SQL server, there are SIDs present for all the DB roles; however, in BBF there is no need to a special SID and hence it is NULL

SQL Server
```
On master - 

1> select name,sid from sys.database_principals;
2> go
name sid
---- ---
public 0x010500000000000904000000FB01993B66F9C34DBD9B2735F4CC0C93
dbo 0x01
guest 0x00
INFORMATION_SCHEMA NULL
sys NULL
##MS_PolicyEventProcessingLogin## 0xF0F3A466D41B4149BFB4467225FE7B03
##MS_AgentSigningCertificate## 0x010600000000000901000000750120FC99725EEE80BF9237AC59F57982426F97
testUser 0xE6A9BDDB1A52F64D9F0390271D5871CB
userof_loginWoSecurityAdmin 0x5FB9CB1D37416643BD74B421FCD0CCA5 -----> deleted login, won't find a mapping in server_principals --- i.e orphaned user
userof_loginWithSecurityAdmin 0x1D208BE96521CE438E6DE67101F00781
userof_logintest 0x608B2654E46ED447B18DA78072FEA960
db_owner 0x01050000000000090400000000000000000000000000000000400000
db_accessadmin 0x01050000000000090400000000000000000000000000000001400000
db_securityadmin 0x01050000000000090400000000000000000000000000000002400000
db_ddladmin 0x01050000000000090400000000000000000000000000000003400000
db_backupoperator 0x01050000000000090400000000000000000000000000000005400000
db_datareader 0x01050000000000090400000000000000000000000000000006400000
db_datawriter 0x01050000000000090400000000000000000000000000000007400000
db_denydatareader 0x01050000000000090400000000000000000000000000000008400000
db_denydatawriter 0x01050000000000090400000000000000000000000000000009400000




On new_db (different owner - loginWithSecurityAdmin)

1> use new_db
2> go
Changed database context to 'new_db'.
1> select name,sid from sys.database_principals;
2> go
name sid
---- ---
public 0x01050000000000090400000083741B006749C04BA943C02702F2A762
dbo 0x1D208BE96521CE438E6DE67101F00781 --------> loginWithSecurityAdmin
guest 0x00
INFORMATION_SCHEMA NULL
sys NULL
db_owner 0x01050000000000090400000000000000000000000000000000400000
db_accessadmin 0x01050000000000090400000000000000000000000000000001400000
db_securityadmin 0x01050000000000090400000000000000000000000000000002400000
db_ddladmin 0x01050000000000090400000000000000000000000000000003400000
db_backupoperator 0x01050000000000090400000000000000000000000000000005400000
db_datareader 0x01050000000000090400000000000000000000000000000006400000
db_datawriter 0x01050000000000090400000000000000000000000000000007400000
db_denydatareader 0x01050000000000090400000000000000000000000000000008400000
db_denydatawriter 0x01050000000000090400000000000000000000000000000009400000

```


Babelfish
```
On master

1> select name, sid from sys.database_principals;
2> go
name                                                                                                                             sid                                                                                                                                                                         
-------------------------------------------------------------------------------------------------------------------------------- ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
db_owner                                                                                                                         NULL                                                                                                                                                                        
dbo                                                                                                                              0x00004000                                                                                                                                                                  
db_datareader                                                                                                                    NULL                                                                                                                                                                        
db_datawriter                                                                                                                    NULL                                                                                                                                                                        
db_accessadmin                                                                                                                   NULL                                                                                                                                                                        
db_securityadmin                                                                                                                 NULL                                                                                                                                                                        
db_ddladmin                                                                                                                      NULL                                                                                                                                                                        
guest                                                                                                                            0x00000000                                                                                                                                                                  
userof_testlogin                                                                                                                 0x00014DF5                                                                                                                                                                  
userof_testlogin_tbd_orphaned                                                                                                    NULL                                                                                                                                                                        
public                                                                                                                           0x00000000                                                                                                                                                                  
sys                                                                                                                              0x00000000                                                                                                                                                                  
INFORMATION_SCHEMA                                                                                                               0x00000000   




On new_db

1> use testdb
2> go
Changed database context to 'testdb'.
1> select name,sid from sys.database_principals;
2> go
name                                                                                                                             sid                                                                                                                                                                         
-------------------------------------------------------------------------------------------------------------------------------- ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
guest                                                                                                                            0x00000000                                                                                                                                                                  
db_ddladmin                                                                                                                      NULL                                                                                                                                                                        
db_securityadmin                                                                                                                 NULL                                                                                                                                                                        
db_accessadmin                                                                                                                   NULL                                                                                                                                                                        
db_datawriter                                                                                                                    NULL                                                                                                                                                                        
db_datareader                                                                                                                    NULL                                                                                                                                                                        
dbo                                                                                                                              0x00014DF5                                                                                                                                                                  
db_owner                                                                                                                         NULL                                                                                                                                                                        
public                                                                                                                           0x00000000                                                                                                                                                                  
sys                                                                                                                              0x00000000                                                                                                                                                                  
INFORMATION_SCHEMA                                                                                                               0x00000000                                                                                                                                                                                                                                                                                                       

```

### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

ayushdsh committed:
Rectified sys.database_principals view which was missing SID

sys.database_principals.sid acts as a foreign key and links to sys.server_principals.sid
This is how SQL users are linked to their respective logins. Certain database_principals like db roles will not have a corresponding mapping in server_principals, hence they have a SID which is generated by default by sql server at creation.
In this view, we want to map the users (the defaults like dbo as well) to the server_principals. The owner of the database is the login which is linked to 'dbo'. Hence, we will be mapping the owner of the current db to 'dbo'. Other user-defined logins are already part of the view currently.

Task: BABEL-5789

ayushdsh committed:
Added testcases for database_principals view

shalinilohia50 commented on contrib/babelfishpg_tsql/sql/ownership.sql:
```diff
@@ -425,7 +425,13 @@ CAST(Ext.default_schema_name AS SYS.SYSNAME) AS default_schema_name,
 CAST(Ext.modify_date AS SYS.DATETIME) AS modify_date,
 CAST(Ext.owning_principal_id AS INT) AS owning_principal_id,
-CAST(CAST(Base2.oid AS INT) AS SYS.VARBINARY(85)) AS SID,
+CAST(
```
These sql changes should also go to upgrade script. https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/BABEL_5_X_DEV/contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--5.2.0--5.3.0.sql

ayushdsh replied:
You mean also to upgrade script right? It will also stay here.

shalinilohia50 replied:
Yes

ayushdsh committed:
Added sys.database_principals to upgrade scripts

ayushdsh committed:
Rectified testcase

shalinilohia50 commented on test/JDBC/expected/sys_database_principals-vu-verify.out:
```diff
@@ -0,0 +1,63 @@
+~~END~~
+
+
+use sys_database_principals_db_different_owner
```
Write "GO" after USE database to remove this warning.

> use sys_database_principals_db_different_owner
> GO

shalinilohia50 commented on test/JDBC/input/views/sys_database_principals-vu-verify.sql:
```diff
@@ -0,0 +1,16 @@
+select suser_sname(sid) from database_principals where name = 'dbo';
```
This test should go to `latest/schedule` to run in other frameworks like mvu/dump-restore.

ayushdsh committed:
Added test file to upgrade schedule and rectified verify script to
remove warning

ayushdsh committed:
Changed test to use custom db to avoid other tests interferring

ayushdsh committed:
Added sys_database_principals to be ignored in single_db mode tests
because these tests need to create multiple dbs

ayushdsh committed:
Altered test behaviour to add sysadmin logins in verify because of tests
interference

shalinilohia50 commented:
Changes look good to me. Please fix the conflicts.

ayushdsh commented:
> Changes look good to me. Please fix the conflicts.

@shalinilohia50 yup. Waiting for PR-3733 to get merged.

ayushdsh committed:
Merge branch 'BABEL_5_X_DEV' into babel-5789

ayushdsh committed:
Rectified sys.database_principals view which was missing SID

sys.database_principals.sid acts as a foreign key and links to sys.server_principals.sid
This is how SQL users are linked to their respective logins. Certain database_principals like db roles will not have a corresponding mapping in server_principals, hence they have a SID which is generated by default by sql server at creation.
In this view, we want to map the users (the defaults like dbo as well) to the server_principals. The owner of the database is the login which is linked to 'dbo'. Hence, we will be mapping the owner of the current db to 'dbo'. Other user-defined logins are already part of the view currently.

Task: BABEL-5789

Signed-off-by: Ayush Shah <ayushdsh@amazon.com>

ayushdsh committed:
Rectified sys.database_principals view which was missing SID

sys.database_principals.sid acts as a foreign key and links to sys.server_principals.sid
This is how SQL users are linked to their respective logins. Certain database_principals like db roles will not have a corresponding mapping in server_principals, hence they have a SID which is generated by default by sql server at creation.
In this view, we want to map the users (the defaults like dbo as well) to the server_principals. The owner of the database is the login which is linked to 'dbo'. Hence, we will be mapping the owner of the current db to 'dbo'. Other user-defined logins are already part of the view currently.

Task: BABEL-5789

Signed-off-by: Ayush Shah <ayushdsh@amazon.com>

ayushdsh committed:
Merge branch 'babelfish-for-postgresql:BABEL_5_X_DEV' into babel-5789

anju15bharti commented on contrib/babelfishpg_tsql/sql/ownership.sql:
```diff
@@ -425,7 +425,13 @@ CAST(Ext.default_schema_name AS SYS.SYSNAME) AS default_schema_name,
 CAST(Ext.owning_principal_id AS INT) AS owning_principal_id,
-CAST(CAST(Base2.oid AS INT) AS SYS.VARBINARY(85)) AS SID,
+CAST(
+  CASE 
+    WHEN Ext.orig_username = 'dbo' THEN CAST(Base3.oid AS INT)
+    WHEN Ext.orig_username = 'guest' THEN 0
+    ELSE CAST(Base2.oid AS INT)
+  END 
+AS SYS.VARBINARY(85)) AS SID,
```
We show same value for sid also in user_token view. Incorporate this changes in that view as well.

sys.user_token view should also have the similar behaviour.

SQL server
```
-- On master - 
1> select suser_sname(sid) from sys.user_token where name = 'dbo';
2> go
 
-
sa

(1 rows affected)

-- On new_db (different owner - l2)
[abanju@dev-dsk-abanju-1c-ea0b53a0]~/oss% tsqlcon -U l2 -d tk
1> select suser_sname(sid) from sys.user_token where name = 'dbo';
2> go
 
-
l2

(1 rows
```

anju15bharti commented on test/JDBC/expected/sys_database_principals-vu-verify.out:
```diff
@@ -0,0 +1,118 @@
+GO
+
+-- create a login which is a member of sysadmin server role
+create login sys_database_principals_vu_login_with_sysadmin with password = '12345678'
```
Create login in prepare script.

ayushdsh replied:
This login has sysadmin permissions, hence it interferes with other tests during version upgrades. Hence, created it here.

anju15bharti replied:
Make it sysadmin in verify file but create the login in prepare script.

anju15bharti commented on test/JDBC/expected/sys_database_principals-vu-verify.out:
```diff
@@ -0,0 +1,118 @@
+GO
+
+-- change db owner to the new sysadmin login
+alter authorization on database::sys_database_principals_db_different_owner to sys_database_principals_vu_login_with_sysadmin
```
Let's make unprivileged login owner of database and connect with that login to get sid of dbo.

anju15bharti commented on test/JDBC/expected/sys_database_principals-vu-verify.out:
```diff
@@ -0,0 +1,118 @@
+~~END~~
+
+
+select suser_sname(sid) from database_principals where name = 'sys_database_principals_vu_orphaned_user' 
+OR name in ('db_owner', 'db_securityadmin', 'db_datawriter', 'db_accessadmin', 'db_datareader', 'db_ddladmin')
```
let's include `sys_database_principals_vu_orphaned_user` inside in clause.

anju15bharti commented on test/JDBC/expected/sys_database_principals-vu-verify.out:
```diff
@@ -0,0 +1,118 @@
+select name, suser_sname(sid) from database_principals order by name;
+GO
+~~START~~
+varchar#!#nvarchar
+db_accessadmin#!#<NULL>
+db_datareader#!#<NULL>
+db_datawriter#!#<NULL>
+db_ddladmin#!#<NULL>
+db_owner#!#<NULL>
+db_securityadmin#!#<NULL>
+dbo#!#jdbc_user
+guest#!#<NULL>
+INFORMATION_SCHEMA#!#<NULL>
+public#!#<NULL>
+sys#!#<NULL>
```
So is there no need to modify the SID value for database roles, nor should a dummy SID value be added to match SQL Server behavior?

ayushdsh replied:
I think a NULL is better than a random SID. SIDs might be used by users to look up things.

shalinilohia50 replied:
The SID of these fixed roles are fixed in all the databases. We should hardcode these values for the database roles to match TSQL behaviour. 

```
1> select name,sid from sys.database_principals;
2> go
db_owner 0x01050000000000090400000000000000000000000000000000400000
db_accessadmin 0x01050000000000090400000000000000000000000000000001400000
db_securityadmin 0x01050000000000090400000000000000000000000000000002400000
db_ddladmin 0x01050000000000090400000000000000000000000000000003400000
db_backupoperator 0x01050000000000090400000000000000000000000000000005400000  --> not supported yet
db_datareader 0x01050000000000090400000000000000000000000000000006400000
db_datawriter 0x01050000000000090400000000000000000000000000000007400000
db_denydatareader 0x01050000000000090400000000000000000000000000000008400000 --> not supported yet
db_denydatawriter 0x01050000000000090400000000000000000000000000000009400000 --> not supported yet
```

anju15bharti replied:
Yes Let's give some value of sid for fixed db roles rather than keeping it NULL.

ayushdsh committed:
Rectified sys.database_principals view which was missing SID

sys.database_principals.sid acts as a foreign key and links to sys.server_principals.sid
This is how SQL users are linked to their respective logins. Certain database_principals like db roles will not have a corresponding mapping in server_principals, hence they have a SID which is generated by default by sql server at creation.
In this view, we want to map the users (the defaults like dbo as well) to the server_principals. The owner of the database is the login which is linked to 'dbo'. Hence, we will be mapping the owner of the current db to 'dbo'. Other user-defined logins are already part of the view currently.

Task: BABEL-5789

Signed-off-by: Ayush Shah <ayushdsh@amazon.com>

ayushdsh committed:
Rectified sys.database_principals view which was missing SID

sys.database_principals.sid acts as a foreign key and links to sys.server_principals.sid
This is how SQL users are linked to their respective logins. Certain database_principals like db roles will not have a corresponding mapping in server_principals, hence they have a SID which is generated by default by sql server at creation.
In this view, we want to map the users (the defaults like dbo as well) to the server_principals. The owner of the database is the login which is linked to 'dbo'. Hence, we will be mapping the owner of the current db to 'dbo'. Other user-defined logins are already part of the view currently.

Task: BABEL-5789

Signed-off-by: Ayush Shah <ayushdsh@amazon.com>

ayushdsh committed:
Rectified sys.database_principals view which was missing SID

sys.database_principals.sid acts as a foreign key and links to sys.server_principals.sid
This is how SQL users are linked to their respective logins. Certain database_principals like db roles will not have a corresponding mapping in server_principals, hence they have a SID which is generated by default by sql server at creation.
In this view, we want to map the users (the defaults like dbo as well) to the server_principals. The owner of the database is the login which is linked to 'dbo'. Hence, we will be mapping the owner of the current db to 'dbo'. Other user-defined logins are already part of the view currently.

Task: BABEL-5789

Signed-off-by: Ayush Shah <ayushdsh@amazon.com>

ayushdsh committed:
Rectified sys.database_principals view which was missing SID

sys.database_principals.sid acts as a foreign key and links to sys.server_principals.sid
This is how SQL users are linked to their respective logins. Certain database_principals like db roles will not have a corresponding mapping in server_principals, hence they have a SID which is generated by default by sql server at creation.
In this view, we want to map the users (the defaults like dbo as well) to the server_principals. The owner of the database is the login which is linked to 'dbo'. Hence, we will be mapping the owner of the current db to 'dbo'. Other user-defined logins are already part of the view currently.

Task: BABEL-5789

Signed-off-by: Ayush Shah <ayushdsh@amazon.com>

ayushdsh committed:
Rectified sys.database_principals view which was missing SID

sys.database_principals.sid acts as a foreign key and links to sys.server_principals.sid
This is how SQL users are linked to their respective logins. Certain database_principals like db roles will not have a corresponding mapping in server_principals, hence they have a SID which is generated by default by sql server at creation.
In this view, we want to map the users (the defaults like dbo as well) to the server_principals. The owner of the database is the login which is linked to 'dbo'. Hence, we will be mapping the owner of the current db to 'dbo'. Other user-defined logins are already part of the view currently.

Task: BABEL-5789

Signed-off-by: Ayush Shah <ayushdsh@amazon.com>

anju15bharti commented on test/JDBC/expected/user_token-dep-vu-verify.out:
```diff
@@ -0,0 +1,42 @@
+~~END~~
+
+
+-- tsql user=user_token_login_with_dbrole password=12345678
```
Add testcases
*  where user is member of db_owner and then query sid from user_token after connecting with the user.
* login is database owner, connect with that login and query same.

ayushdsh committed:
Rectified sys.database_principals view which was missing SID

sys.database_principals.sid acts as a foreign key and links to sys.server_principals.sid
This is how SQL users are linked to their respective logins. Certain database_principals like db roles will not have a corresponding mapping in server_principals, hence they have a SID which is generated by default by sql server at creation.
In this view, we want to map the users (the defaults like dbo as well) to the server_principals. The owner of the database is the login which is linked to 'dbo'. Hence, we will be mapping the owner of the current db to 'dbo'. Other user-defined logins are already part of the view currently.

Task: BABEL-5789

Signed-off-by: Ayush Shah <ayushdsh@amazon.com>

ayushdsh committed:
Rectified sys.database_principals view which was missing SID

sys.database_principals.sid acts as a foreign key and links to sys.server_principals.sid
This is how SQL users are linked to their respective logins. Certain database_principals like db roles will not have a corresponding mapping in server_principals, hence they have a SID which is generated by default by sql server at creation.
In this view, we want to map the users (the defaults like dbo as well) to the server_principals. The owner of the database is the login which is linked to 'dbo'. Hence, we will be mapping the owner of the current db to 'dbo'. Other user-defined logins are already part of the view currently.

Task: BABEL-5789

Signed-off-by: Ayush Shah <ayushdsh@amazon.com>

ayushdsh committed:
Rectified sys.database_principals view which was missing SID

sys.database_principals.sid acts as a foreign key and links to sys.server_principals.sid
This is how SQL users are linked to their respective logins. Certain database_principals like db roles will not have a corresponding mapping in server_principals, hence they have a SID which is generated by default by sql server at creation.
In this view, we want to map the users (the defaults like dbo as well) to the server_principals. The owner of the database is the login which is linked to 'dbo'. Hence, we will be mapping the owner of the current db to 'dbo'. Other user-defined logins are already part of the view currently.

Task: BABEL-5789

Signed-off-by: Ayush Shah <ayushdsh@amazon.com>

ayushdsh committed:
Rectified sys.database_principals view which was missing SID

sys.database_principals.sid acts as a foreign key and links to sys.server_principals.sid
This is how SQL users are linked to their respective logins. Certain database_principals like db roles will not have a corresponding mapping in server_principals, hence they have a SID which is generated by default by sql server at creation.
In this view, we want to map the users (the defaults like dbo as well) to the server_principals. The owner of the database is the login which is linked to 'dbo'. Hence, we will be mapping the owner of the current db to 'dbo'. Other user-defined logins are already part of the view currently.

Task: BABEL-5789

Signed-off-by: Ayush Shah <ayushdsh@amazon.com>

shalinilohia50 commented on test/JDBC/expected/sys_database_principals-vu-verify.out:
```diff
@@ -0,0 +1,173 @@
+~~END~~
+
+
+select suser_sname(sid) from database_principals 
```
This test doesn't add much value since we can't rely on fixed db role SID to find its name[SID being different from the OID]. Lets extend this to show the constant SID value as well.

anju15bharti commented on test/JDBC/expected/user_token-dep-vu-verify.out:
```diff
@@ -0,0 +1,126 @@
+
+-- terminate-tsql-conn user=user_token_login_with_dbrole password=12345678
+
+-- tsql user=user_token_login_with_db_owner password=12345678
+select * from user_token_vu order by name;
+GO
+~~START~~
+nvarchar#!#nvarchar#!#nvarchar#!#nvarchar
+db_accessadmin#!#<NULL>#!#ROLE#!#GRANT OR DENY
+db_datareader#!#<NULL>#!#ROLE#!#GRANT OR DENY
+db_datawriter#!#<NULL>#!#ROLE#!#GRANT OR DENY
+db_ddladmin#!#<NULL>#!#ROLE#!#GRANT OR DENY
+db_owner#!#<NULL>#!#ROLE#!#GRANT OR DENY
+db_securityadmin#!#<NULL>#!#ROLE#!#GRANT OR DENY
+public#!#<NULL>#!#ROLE#!#GRANT OR DENY
+testrole#!#<NULL>#!#ROLE#!#GRANT OR DENY
+u_user_token_login_with_db_owner#!#user_token_login_with_db_owner#!#SQL USER#!#GRANT OR DENY
+~~END~~
+
+select * from dbo.user_token_fixed_roles;
+GO
+~~START~~
+nvarchar#!#varbinary
+db_accessadmin#!#01050000000000090400000000000000000000000000000001400000
+db_datareader#!#01050000000000090400000000000000000000000000000006400000
+db_datawriter#!#01050000000000090400000000000000000000000000000007400000
+db_ddladmin#!#01050000000000090400000000000000000000000000000003400000
+db_owner#!#01050000000000090400000000000000000000000000000000400000
+db_securityadmin#!#01050000000000090400000000000000000000000000000002400000
+~~END~~
```
I don’t believe the user is a member of all the fixed database roles. From what I can see, the user is only a member of the db_owner role, so only the db_owner entry should appear in the results, not all the other fixed database roles.

I think the bug is with user being member of db_owner as it ideally has all other db roles priv but for this view, it should only show entry for those roles of which the current user is explicitly member of.

We can either fix this behavior or, alternatively, create a separate Jira(if not already there) to address it and n the meantime, we should ensure that this PR includes only the correct test cases, such as making the user a member of all database roles to verify the output in this view.

ayushdsh replied:
Yes. This is an existing issue and happens because we check pg_has_role, and a user being member of db_owner becomes member of all other roles. I think we should create another JIRA for this; but I think we should keep this test; once that fix is in, this should fail.

anju15bharti replied:
Once this issue is fixed, our test coverage will be reduced, as it will only show the db_owner entry. However, we want to ensure that all database roles are displayed, so we need to adjust the test cases accordingly to reflect the correct behavior.

shalinilohia50 replied:
I would suggest lets keep this test since this is an existing issue. Create a JIRA with this test information, so that we know this test was expected to fail - Add that in the JIRA description.

ayushdsh committed:
Rectified sys.database_principals view which was missing SID

sys.database_principals.sid acts as a foreign key and links to sys.server_principals.sid
This is how SQL users are linked to their respective logins. Certain database_principals like db roles will not have a corresponding mapping in server_principals, hence they have a SID which is generated by default by sql server at creation.
In this view, we want to map the users (the defaults like dbo as well) to the server_principals. The owner of the database is the login which is linked to 'dbo'. Hence, we will be mapping the owner of the current db to 'dbo'. Other user-defined logins are already part of the view currently.

Task: BABEL-5789

Signed-off-by: Ayush Shah <ayushdsh@amazon.com>

ayushdsh committed:
Rectified sys.database_principals view which was missing SID

sys.database_principals.sid acts as a foreign key and links to sys.server_principals.sid
This is how SQL users are linked to their respective logins. Certain database_principals like db roles will not have a corresponding mapping in server_principals, hence they have a SID which is generated by default by sql server at creation.
In this view, we want to map the users (the defaults like dbo as well) to the server_principals. The owner of the database is the login which is linked to 'dbo'. Hence, we will be mapping the owner of the current db to 'dbo'. Other user-defined logins are already part of the view currently.

Task: BABEL-5789

Signed-off-by: Ayush Shah <ayushdsh@amazon.com>

shalinilohia50 reviewed (approved):
LGTM!

anju15bharti reviewed (approved):
LGTM, Just update the description to include information about sys.user_token change as well.

jsudrik reviewed (approved):
Approved.
